Add (cancelable) ToastAddEvent for use in ToastGui.add

Roundaround opened the pull request:
Gives modders the ability to hook into the ToastGui's toast queue. To prevent the toast from being added, cancel the event. To change the toast, modify it directly or replace it with event.setToast(IToast).

Roundaround committed:
Add (cancelable) ToastAddEvent for use in ToastGui.add

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=7821) <br/>All committers have signed the CLA.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/event/ToastAddEvent.java:
```diff
@@ -0,0 +1,21 @@
+package net.minecraftforge.client.event;
```
Missing license header.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/event/ToastAddEvent.java:
```diff
@@ -0,0 +1,21 @@
+import net.minecraftforge.eventbus.api.Event;
+
+@Cancelable
+public class ToastAddEvent extends Event {
```
Incorrect formatting, forge uses new line braces.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/event/ToastAddEvent.java:
```diff
@@ -0,0 +1,21 @@
+@Cancelable
+public class ToastAddEvent extends Event {
+    private IToast toast;
+    public ToastAddEvent(IToast toast) {
```
New line braces.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/event/ToastAddEvent.java:
```diff
@@ -0,0 +1,21 @@
+
+@Cancelable
+public class ToastAddEvent extends Event {
+    private IToast toast;
```
Separate fields from constructor with empty line.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/event/ToastAddEvent.java:
```diff
@@ -0,0 +1,21 @@
+        setToast(toast);
+    }
+
+    public IToast getToast() {
```
New line braces.

ChampionAsh5357 commented on src/main/java/net/minecraftforge/client/event/ToastAddEvent.java:
```diff
@@ -0,0 +1,21 @@
+        return toast;
+    }
+
+    public void setToast(IToast toast) {
```
New line braces.

Roundaround committed:
Save event.getToast() back to existing parameter to avoid needing to remove line

ChampionAsh5357 reviewed (commented):
Bunch of things that would probably need to be addressed.
- There is no test case. How do we know this works rather than just some assumption that it does.
- There is no documentation. How do I use this event? What does it do?
- This will break existing Minecraft behavior because it doesn't replace all cases of the toast. e.g. `Tutorial#addTimedToast`. This will break all tutorial steps if replaced.

ChampionAsh5357 commented on patches/minecraft/net/minecraft/client/gui/toasts/ToastGui.java.patch:
```diff
@@ -1,13 +1,12 @@
 +      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) return;
-+      this.field_191792_h.add(event.getToast());
++      p_192988_1_ = event.getToast();
+       this.field_191792_h.add(p_192988_1_);
```
Just added another line, revert to what you previously had.

Roundaround committed:
Addressed review comments

Roundaround committed:
Added example/test mod demonstrating the event usage

Roundaround committed:
Added missing Forge license header

Roundaround committed:
Removed ability to replace the toast instance within the event to avoid breaking situations like Tutorial#addTimedToast

Roundaround commented:
I've formatted the files as requested, added JavaDoc comments to the event, and written a text/example mod demonstrating the event usage.

As for your third point about situations like `Tutorial#addTimedToast`, I've removed the ability to set the toast in the event (and therefore made it final), so that references to the passed toast remain valid.

Roundaround commented:
Sorry, accidentally closed when I attempted to rename the branch. Apparently you can rename the target branch and PRs will update, but the same cannot be said for the source branch.

Roundaround reviewed (commented):
Responded to feedback. Not sure if I have to mark it reviewed to bounce it back? I've never used GitHub's code review system before

TheCurle commented on src/test/java/net/minecraftforge/debug/client/ToastAddTest.java:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    @OnlyIn(Dist.CLIENT)
```
This is poor practice, and in actuality won't help prevent any issues.. this should be removed from the test mod.

TheCurle reviewed (changes requested):
Seems good, and if you can fix the issue quickly we can merge this with the next batch of PRs.

sciwhiz12 commented:
Hello! @Roundaround, please leave a comment whether you wish to continue this PR and update it to the latest active version of 1.18 (or leave an explanation why this PR's changes are not applicable to 1.18). If we do not hear back from you after some time, this PR will be closed for inactivity.
If anyone else wishes to continue the work of this PR, please create a PR with a link to this PR for reference. 
Thank you for your contributions. 👋
